### PR TITLE
docs: Document the new signature interface for providers

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by levitte** [comment]
Please review / comment, even though it's still a draft

File: doc/man7/provider-signature.pod [hunk 1]
```diff
@@ -22,17 +22,35 @@ provider-signature - The signature library E<lt>-E<gt> provider functions
...
  int OSSL_FUNC_signature_verify_init(void *ctx, void *provkey,
                                      const OSSL_PARAM params[]);
  int OSSL_FUNC_signature_verify(void *ctx, const unsigned char *sig, size_t siglen,
                                 const unsigned char *tbs, size_t tbslen);
+ int OSSL_FUNC_signature_verify_message_init(void *ctx, void *provkey,
+                                             const OSSL_PARAM params[]);
+ int OSSL_FUNC_signature_verify_message_update(void *ctx, const unsigned char *in,
+                                               size_t inlen);
+ /*
+  * OSSL_FUNC_signature_verify_final requires that the signature to be verified
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:50]: I assume you mean `OSSL_FUNC_signature_verify_message_final`?

File: doc/man7/provider-signature.pod [hunk 2]
```diff
@@ -22,17 +22,35 @@ provider-signature - The signature library E<lt>-E<gt> provider functions
...
+  * against is specified via an OSSL_PARAM.
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:51]: Which function is used to supply the `OSSL_PARAM`?

File: doc/man7/provider-signature.pod [hunk 3]
```diff
@@ -216,6 +249,38 @@ The length of the signature should be written to I<*siglen>.
 If I<sig> is NULL then the maximum length of the signature should be written to
 I<*siglen>.
 
+=head2 Message Signing Functions
+
+These functions are suitable to algorithms that implement algorithms that
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:254]: algorithms that implement algorithms??

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/provider-signature.pod (hunk 2 above)
Comment (levitte) [reply doc/man7/provider-signature.pod:51]: Expanded that comment a bit

**Comment by levitte** [comment]
I have sufficiently massaged this manual for my taste.  It's ready for
final reviews.

File: doc/man7/provider-signature.pod [hunk 4]
```diff
@@ -216,6 +250,38 @@ The length of the signature should be written to I<*siglen>.
 If I<sig> is NULL then the maximum length of the signature should be written to
 I<*siglen>.
 
+=head2 Message Signing Functions
+
+These functions are suitable for providers that implement algorithms that
+accumulates a full message and signs the result of that accumulation, such as
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:256]: typo: accumulate

File: doc/man7/provider-signature.pod [hunk 5]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
 The signature is pointed to by the I<sig> parameter which is I<siglen> bytes
 long.
 
+=head2 Message Verify Functions
+
+These functions are suitable to algorithms that implement algorithms that
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:305]: We sill have algorithms that implement algorithms :-)

File: doc/man7/provider-signature.pod [hunk 6]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+accumulates a full message and signs the result of that accumulation, such as
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:306]: These are the verify functions - so why are we talking about signing here?

File: doc/man7/provider-signature.pod [hunk 7]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+RSA-SHA256.
+
+OSSL_FUNC_signature_verify_message_init() initialises a context for signing a
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:309]: s/signing/verifying/

File: doc/man7/provider-signature.pod [hunk 8]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+The I<params>, if not NULL, should be set on the context in a manner similar to
+using OSSL_FUNC_signature_set_ctx_params().
+The key object should have been previously generated, loaded or imported into
+the provider using the key management (OSSL_OP_KEYMGMT) operation (see
+provider-keymgmt(7)>.
+
+OSSL_FUNC_signature_verify_message_update() gathers the data pointed at by
+I<in>, which is I<inlen> bytes long.
+
+OSSL_FUNC_signature_verify_message_final() performs the actual signing on the
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:321]: s/signing on/verifying of/

File: doc/man7/provider-signature.pod [hunk 9]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+data that was gathered with OSSL_FUNC_signature_verify_message_update().  The
+signature itself must have been passed through the "signature"
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:323]: Is this supposed to say 'passed through the "signature" parameter'

File: doc/man7/provider-signature.pod [hunk 10]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+(B<OSSL_SIGNATURE_PARAM_SIGNATURE>) L<Signature parameter|/Signature parameters>
+before this function is called.
+
+OSSL_FUNC_signature_verify() can be used for one-shot signature calls.  In that
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:327]: s/signature/verification/

File: doc/man7/provider-signature.pod [hunk 11]
```diff
@@ -234,6 +300,34 @@ is I<tbslen> bytes long.
...
+case, I<tbs> is expected to be the whole message to be signed, I<tbslen> bytes
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:328]: Do we normally use `tbs` for verify operations?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man7/provider-signature.pod (hunk 11 above)
Comment (levitte) [reply doc/man7/provider-signature.pod:328]: Referring back to 3.3.2:

https://github.com/openssl/openssl/blob/fb7fab9fa6f4869eaa8fbb97e0d593159f03ffe4/doc/man3/EVP_PKEY_verify.pod#L14-L16
https://github.com/openssl/openssl/blob/fb7fab9fa6f4869eaa8fbb97e0d593159f03ffe4/doc/man7/provider-signature.pod#L34-L35

File: doc/man7/provider-signature.pod (hunk 6 above)
Comment (levitte) [reply doc/man7/provider-signature.pod:306]: Changed

File: doc/man7/provider-signature.pod (hunk 5 above)
Comment (levitte) [reply doc/man7/provider-signature.pod:305]: Fixed

File: doc/man7/provider-signature.pod (hunk 9 above)
Comment (levitte) [reply doc/man7/provider-signature.pod:323]: It does, on the next line (with a reference to the Signature parameters section)

File: doc/man7/provider-signature.pod [hunk 12]
```diff
@@ -234,6 +300,35 @@ is I<tbslen> bytes long.
...
+I<in>, which is I<inlen> bytes long.
+
+OSSL_FUNC_signature_verify_message_final() performs the actual verification on
+the data that was gathered with OSSL_FUNC_signature_verify_message_update().
+The signature itself must have been passed through the "signature"
+(B<OSSL_SIGNATURE_PARAM_SIGNATURE>) L<Signature parameter|/Signature parameters>
+before this function is called.
+
+OSSL_FUNC_signature_verify() can be used for one-shot verification calls.  In
+that case, I<tbs> is expected to be the whole message to be signed, I<tbslen>
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:329]: s/to be signed/to be verified/

File: doc/man7/provider-signature.pod [hunk 13]
```diff
@@ -234,6 +300,35 @@ is I<tbslen> bytes long.
 The signature is pointed to by the I<sig> parameter which is I<siglen> bytes
 long.
 
+=head2 Message Verify Functions
+
+These functions are suitable for providers that implement algorithms that
+accumulate a full message and verifies a signature against the result
+of that accumulation, such as RSA-SHA256.
+
+OSSL_FUNC_signature_verify_message_init() initialises a context for verify a
```
Comment (mattcaswell) [inline doc/man7/provider-signature.pod:309]: s/verify/verifying/

**Review by mattcaswell — APPROVED** [review]
Approved subject to the final couple of nits being fixed.

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — DISMISSED** [review]

**Review by shahsb — CHANGES_REQUESTED** [review]
Gave minor comments. Other content LGTM. Thanks for making the changes!

File: doc/man7/provider-signature.pod [hunk 14]
```diff
@@ -216,6 +250,38 @@ The length of the signature should be written to I<*siglen>.
 If I<sig> is NULL then the maximum length of the signature should be written to
 I<*siglen>.
 
+=head2 Message Signing Functions
+
+These functions are suitable for providers that implement algorithms that
+accumulate a full message and signs the result of that accumulation, such as
+RSA-SHA256.
+
+OSSL_FUNC_signature_sign_message_init() initialises a context for signing a
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-signature.pod:259]: "initialises" is British English spelling

File: doc/man7/provider-signature.pod [hunk 15]
```diff
@@ -216,6 +250,38 @@ The length of the signature should be written to I<*siglen>.
...
+message given a provider side signature context in the I<ctx> parameter, and a
+pointer to a provider key object in the I<provkey> parameter.
+The I<params>, if not NULL, should be set on the context in a manner similar to
+using OSSL_FUNC_signature_set_ctx_params().
+The key object should have been previously generated, loaded or imported into
+the provider using the key management (OSSL_OP_KEYMGMT) operation (see
+provider-keymgmt(7)>.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-signature.pod:266]: Nice catch!  The correction isn't quite right, but thanks for bringing it to attention!

File: doc/man7/provider-signature.pod [hunk 16]
```diff
@@ -216,6 +250,38 @@ The length of the signature should be written to I<*siglen>.
...
 I<*siglen>.
 
+=head2 Message Signing Functions
+
+These functions are suitable for providers that implement algorithms that
+accumulate a full message and signs the result of that accumulation, such as
+RSA-SHA256.
+
+OSSL_FUNC_signature_sign_message_init() initialises a context for signing a
+message given a provider side signature context in the I<ctx> parameter, and a
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-signature.pod:260]: Arguable, as far as I know.  We use 'provider side' all over the place in our source, while 'provider-side', while being present, is very rare.  So I'll stick to the majority

File: doc/man7/provider-signature.pod [hunk 17]
```diff
@@ -216,6 +250,38 @@ The length of the signature should be written to I<*siglen>.
...
+pointer to a provider key object in the I<provkey> parameter.
+The I<params>, if not NULL, should be set on the context in a manner similar to
+using OSSL_FUNC_signature_set_ctx_params().
+The key object should have been previously generated, loaded or imported into
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man7/provider-signature.pod:264]: We're flexible re the Oxford comma.  Here, it gives no real benefit...

**Comment by levitte** [comment]
@t8m, @mattcaswell, re-approve please?

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
IMO no need to reset the merge waiting window.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master and 3.4 branches. Thank you.

**Pull request closed without merging**